File: implement_trie.cpp

// A trie (pronounced as "try") or prefix tree is a tree data structure 
// used to efficiently store and retrieve keys in a dataset of strings. 
// There are various applications of this data structure, such as 
// autocomplete and spellchecker.

// Implement the Trie class:

// Trie() Initializes the trie object.
// void insert(String word) Inserts the string word into the trie.
// boolean search(String word) Returns true if the string word is in the 
// trie (i.e., was inserted before), and false otherwise.
// boolean startsWith(String prefix) Returns true if there is a previously
// inserted string word that has the prefix prefix, and false otherwise.

// This solution works fine on my intel machine but the Leet Code sanitizer
// complains about it. 

#include <iostream>
#include <string>

using namespace std;


class Node
{
    public:
    char letter = '\0';
    Node* children[26] = {}; // restricted to a-z lower. lookup table yo!
    bool is_word = false; 
};

class Trie
{ 
    private: 
    Node* root;

    public: 
        Trie() 
        {
            
        }
        
        void insert(string word) 
        {
            if(word == "") return;
            if(root == nullptr)
            {
                //cout << "instantiating root node" << endl;
                root = new Node(); // root starts as !nullptr
            }

            Node* current = root;
            for(int i = 0; i < word.length(); i++)
            {
                int position = tolower(word[i]) - 'a';
                
                // create child if it doesn't exist
                if(current->children[position] == nullptr)
                {
                    // debug
                    //cout << "child not exist, inserting at: " << position << endl;
                    Node* child = new Node();
                    child->letter = tolower(word[i]);
                    //if(i == word.length() - 1) child->is_word = true;
                    current->children[position] = child;
                }                
                
                // move to that child node!
                current = current->children[position];
            }
            current->is_word = true;
        }
        
        bool search(string word) 
        {
            if(word == "")return false;
            if(root == nullptr) return false;
            bool in_tree = false;
            Node* current = root;
            for(int i=0; i < word.length(); i++)
            {
                int position = word[i] - 'a';
                if(current->children[position] == nullptr) return false;
                if(current->children[position]->is_word){
                    in_tree = true;
                    break;
                }
                current = current->children[position];
            }

            return in_tree;
        }
        
        bool startsWith(string prefix) {
            if(prefix == "")return false; 
            if(root == nullptr) return false;
            Node* current = root;

            for(int i=0; i < prefix.length(); i++)
            {
                int position = prefix[i] - 'a';
                if(current->children[position] == nullptr) return false;
                 
                current = current->children[position];
            } 
            return true;
        }

        void print()
        {  
            print_rec(root, "");
        }

        // recursive print
        void print_rec(Node* root, string word_so_far)
        {
            if(root == nullptr) return;
             
            word_so_far += root->letter;

            if(root->is_word)
            {   
                cout << word_so_far << endl;
            }

            for(int i = 0; i < 26; i++)
            {
                print_rec(root->children[i], word_so_far);
            }
        }

};
    int main()
    {
        Trie* t = new Trie();
        t->insert("apple");
        cout << t->search("apple") << endl;
        cout << t->search("app") << endl;
        cout << t->startsWith("app") << endl;
        t->insert("app");
        cout << t->search("app") << endl;

        t->print();
        //cout << "act: " << t->search("act") << endl;
        //cout << "activate: " << t->search("activate") << endl;

        return 0;
    }

File: implement_trie.py
# Trie in python fuark you LEET CODE for not accepting any of my C++ solutions.

class Node:
    def __init__(self):
        self.is_word = False
        self.children = {}
        self.letter = "" 
        

class Trie:

    root = None

    def __init__(self):
        self.root = Node()

    def insert(self, word: str) -> None:
        if(word == None):
            return

        if(self.root == None):
            self.root = Node()

        word.lower()
        current = self.root

        for c in word:
            if c not in current.children:
                child = Node()
                child.letter = c
                current.children[c] = child
            
            current = current.children[c]

        current.is_word = True
            

         

             
    def search(self, word: str) -> bool:
        if word == None:  
            return False
        if self.root == None: 
            return False 

        current = self.root
        for c in word: 
            # if the current's child doesn't exist, no word!
            if c not in current.children:
                return False
              
            # if none of that occured, just go to the next child
            current = current.children[c]

        return current.is_word
        

    def startsWith(self, prefix: str) -> bool:
        if prefix == None:  
            return False
        if self.root == None: 
            return False

        current = self.root
        for c in prefix:

            # if the current's child doesn't exist, no word!
            if c not in current.children:
                return False
            
            # if none of that occured, just go to the next child
            current = current.children[c]

        return True

    def print_trie(self):
        self.print_recursive(self.root, " ")

    def print_recursive(self, current, word_so_far):

        if current == None: 
            return 

        word_so_far = str(word_so_far) + str(current.letter)

        if current.is_word:
            print(word_so_far) 

        for child in current.children.values(): 
            self.print_recursive(child, word_so_far)
 

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

# ["Trie","insert","search","search","startsWith","insert","search"]
# [[],["apple"],["apple"],["app"],["app"],["app"],["app"]]

def main():
    t = Trie()
    t.insert("apple")
    t.insert("appetite")
    t.insert("derp")
    t.insert("barnacle")

    print(t.search("apple"))
    print(t.search("asdf"))
    print(t.search("apples")) 
    print(t.startsWith("app"))
    print(t.startsWith("azz"))
    
    
    #t.print_trie()

main()